### PR TITLE
Adding 2018-02-01 and 2018-11-01 schemas for Microsoft.Web

### PR DESCRIPTION
**Pull request by pvbchak**
[D:\azure-resource-manager-schemas\generator] executing: autorest-beta.cmd --azureresourceschema --output-folder=C:\Users\pvbchak\AppData\Local\Temp\ntgwupqcmie --tag=all-api-versions --api-version=2018-02-01 --title=none C:\Users\pvbchak\AppData\Local\Temp\schm_azspc\specification\web\resource-manager\readme.md
AutoRest code generation utility [version: 3.0.6119; node: v10.16.3, max-memory: 8192 gb]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      'D:\azure-resource-manager-schemas\generator\node_modules\@autorest\core\dist' (3.0.6130)
   Loading local AutoRest extension '@autorest/azureresourceschema' (D:\azure-resource-manager-schemas\generator\node_modules\@autorest\azureresourceschema)
WARNING: Skipping path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/host/default/{keyType}/{keyName}': Parameter reference {keyType} is not defined as an enum
WARNING: Skipping path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/host/default/{keyType}/{keyName}': Parameter reference {keyType} is not defined as an enum
[670.44 s] Generation Complete

================================================================================================================================
Filename: D:\azure-resource-manager-schemas\schemas\2018-02-01\Microsoft.Web.json
Provider Namespace: Microsoft.Web
API Version: 2018-02-01
Resource Types (Tenant Scope):
- publishingUsers
- sourcecontrols
Resource Types (Resource Group Scope):
- certificates
- hostingEnvironments
- hostingEnvironments/multiRolePools
- hostingEnvironments/workerPools
- serverfarms
- serverfarms/virtualNetworkConnections/gateways
- serverfarms/virtualNetworkConnections/routes
- sites
- sites/config
- sites/deployments
- sites/domainOwnershipIdentifiers
- sites/extensions
- sites/functions
- sites/functions/keys
- sites/hostNameBindings
- sites/hybridconnection
- sites/hybridConnectionNamespaces/relays
- sites/instances/extensions
- sites/migrate
- sites/networkConfig
- sites/premieraddons
- sites/privateAccess
- sites/publicCertificates
- sites/siteextensions
- sites/slots
- sites/slots/config
- sites/slots/deployments
- sites/slots/domainOwnershipIdentifiers
- sites/slots/extensions
- sites/slots/functions
- sites/slots/functions/keys
- sites/slots/hostNameBindings
- sites/slots/hybridconnection
- sites/slots/hybridConnectionNamespaces/relays
- sites/slots/instances/extensions
- sites/slots/networkConfig
- sites/slots/premieraddons
- sites/slots/privateAccess
- sites/slots/publicCertificates
- sites/slots/siteextensions
- sites/slots/sourcecontrols
- sites/slots/virtualNetworkConnections
- sites/slots/virtualNetworkConnections/gateways
- sites/sourcecontrols
- sites/virtualNetworkConnections
- sites/virtualNetworkConnections/gateways
================================================================================================================================
[D:\azure-resource-manager-schemas\generator] executing: autorest-beta.cmd --azureresourceschema --output-folder=C:\Users\pvbchak\AppData\Local\Temp\u7cq8p2ujx --tag=all-api-versions --api-version=2018-11-01 --title=none C:\Users\pvbchak\AppData\Local\Temp\schm_azspc\specification\web\resource-manager\readme.md
AutoRest code generation utility [version: 3.0.6119; node: v10.16.3, max-memory: 8192 gb]
(C) 2018 Microsoft Corporation.
https://aka.ms/autorest
   Loading AutoRest core      'D:\azure-resource-manager-schemas\generator\node_modules\@autorest\core\dist' (3.0.6130)
   Loading local AutoRest extension '@autorest/azureresourceschema' (D:\azure-resource-manager-schemas\generator\node_modules\@autorest\azureresourceschema)
[403.22 s] Generation Complete
================================================================================================================================
Filename: D:\azure-resource-manager-schemas\schemas\2018-11-01\Microsoft.Web.json
Provider Namespace: Microsoft.Web
API Version: 2018-11-01
Resource Types (Resource Group Scope):
- certificates
- sites
- sites/config
- sites/deployments
- sites/domainOwnershipIdentifiers
- sites/extensions
- sites/functions
- sites/hostNameBindings
- sites/hybridconnection
- sites/hybridConnectionNamespaces/relays
- sites/instances/extensions
- sites/migrate
- sites/networkConfig
- sites/premieraddons
- sites/privateAccess
- sites/publicCertificates
- sites/siteextensions
- sites/slots
- sites/slots/config
- sites/slots/deployments
- sites/slots/domainOwnershipIdentifiers
- sites/slots/extensions
- sites/slots/functions
- sites/slots/hostNameBindings
- sites/slots/hybridconnection
- sites/slots/hybridConnectionNamespaces/relays
- sites/slots/instances/extensions
- sites/slots/networkConfig
- sites/slots/premieraddons
- sites/slots/privateAccess
- sites/slots/publicCertificates
- sites/slots/siteextensions
- sites/slots/sourcecontrols
- sites/slots/virtualNetworkConnections
- sites/slots/virtualNetworkConnections/gateways
- sites/sourcecontrols
- sites/virtualNetworkConnections
- sites/virtualNetworkConnections/gateways
================================================================================================================================

### CODE REVIEW

File: schemas/2014-04-01-preview/deploymentTemplate.json [hunk 1]
```diff
@@ -1513,6 +1513,252 @@
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_hybridConnections_authorizationRules" },
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_wcfRelays" },
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_wcfRelays_authorizationRules" },
+                  {
```
Comment (anthony-c-martin) [inline schemas/2014-04-01-preview/deploymentTemplate.json]: [nit] match formatting for the other (previous) lines in this file.

File: schemas/2015-01-01/deploymentTemplate.json [hunk 2]
```diff
@@ -1522,7 +1522,253 @@
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_hybridConnections" },
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_hybridConnections_authorizationRules" },
                   { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_wcfRelays" },
-                  { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_wcfRelays_authorizationRules" }
+                  { "$ref": "https://schema.management.azure.com/schemas/2017-04-01/Microsoft.Relay.json#/resourceDefinitions/namespaces_wcfRelays_authorizationRules" },
```
Comment (anthony-c-martin) [inline schemas/2015-01-01/deploymentTemplate.json]: [nit] match formatting for the other (previous) lines in this file.

File: schemas/2019-04-01/deploymentTemplate.json [hunk 3]
```diff
@@ -1388,7 +1388,253 @@
                   { "$ref": "https://schema.management.azure.com/schemas/2018-03-01/Microsoft.Authorization.json#/resourceDefinitions/policyAssignments" },
                   { "$ref": "https://schema.management.azure.com/schemas/2018-05-01/Microsoft.Authorization.json#/resourceDefinitions/policyAssignments" },
                   { "$ref": "https://schema.management.azure.com/schemas/2019-01-01/Microsoft.Authorization.json#/resourceDefinitions/policyAssignments" },
-                  { "$ref": "https://schema.management.azure.com/schemas/2019-06-01/Microsoft.Authorization.json#/resourceDefinitions/policyAssignments" }
+                  { "$ref": "https://schema.management.azure.com/schemas/2019-06-01/Microsoft.Authorization.json#/resourceDefinitions/policyAssignments" },
+                  {
```
Comment (anthony-c-martin) [inline schemas/2019-04-01/deploymentTemplate.json]: [nit] match formatting for the other (previous) lines in this file.

File: schemas/2018-02-01/Microsoft.Web.json [hunk 4]
```diff
@@ -0,0 +1,7804 @@
+{
+  "id": "https://schema.management.azure.com/schemas/2018-02-01/Microsoft.Web.json#",
```
Comment (anthony-c-martin) [inline schemas/2018-02-01/Microsoft.Web.json]: Please attach the console output when running `npm run generate-single ...` for us to review

**Comment by ruslany** [comment]
The changes look correct to me.

**Comment by naveedaz** [comment]
>           "description": "Name of the certificate."

NIT: Should be "Name of the site".

---
Refers to: schemas/2018-02-01/Microsoft.Web.json:426 in b3b8e5b. [](commit_id = b3b8e5bd45e46259ad41ea42c49ecf912de7414b, deletion_comment = False)

**Comment by anthony-c-martin** [comment]
@pvbchak the piece I'm unsure about (from the logs) is:
```
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/authsettings': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/azurestorageaccounts': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/backup': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/connectionstrings': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/logs': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/metadata': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/pushsettings': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/slotConfigNames': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/web': Duplicate resource definition sites_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/authsettings': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/azurestorageaccounts': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/backup': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/connectionstrings': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/logs': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/metadata': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/pushsettings': Duplicate resource definition sites_slots_config
INFORMATION: Skipping resource type Microsoft.Web/sites/slots/config under path '/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/slots/{slot}/config/web': Duplicate resource definition sites_slots_config
```

This suggests that these resources have been omitted entirely from the schema. I've raised https://github.com/Azure/autorest.azureresourceschema/issues/43 for this. Do you know how these particular resource types have been handled in previous versions of your schemas? If it's not a regression then I think we can live with it - otherwise we may want to fix it first.

**Comment by naveedaz** [comment]
>               "pattern": "^.*/appsettings$"

Seems like we are collapsing all site configs into one schema. There are a whole lot of configs we support. 

config/web
config/appsettings
config/connectiionstrings
config/azurestorageaccounts
config/authsettings
config/pushsettings
config/logs
config/backup
config/virtualNetwork
config/slotConfigNames

-- I believe not released yet
config/configreferences

-- I believe does not need to be exposed in the schema
config/metadata

To support, them all as a single schema site_config we would then need to have a pattern like: 

"^.*/(appsettings|web|connectionstrings|azurestorageaccounts|authsettings|pushsettings|logs|backup|VirtualNetwork|slotConfigNames)"

---
Refers to: schemas/2018-02-01/Microsoft.Web.json:557 in b3b8e5b. [](commit_id = b3b8e5bd45e46259ad41ea42c49ecf912de7414b, deletion_comment = False)

**Comment by naveedaz** [comment]
>             },

This only describes config/appsettings but all the others actually have different responses. 

---
Refers to: schemas/2018-02-01/Microsoft.Web.json:588 in b3b8e5b. [](commit_id = b3b8e5bd45e46259ad41ea42c49ecf912de7414b, deletion_comment = False)

**Comment by anthony-c-martin** [comment]
> > ```
> >           "pattern": "^.*/appsettings$"
> > ```
> 
> Seems like we are collapsing all site configs into one schema. There are a whole lot of configs we support.
> 

@naveedaz Right, that's what I was getting at with this comment https://github.com/Azure/azure-resource-manager-schemas/pull/761#issuecomment-543883742, and the issue I opened https://github.com/Azure/autorest.azureresourceschema/issues/43.

If I'm interpreting your swagger definition correctly, it's actually not enough to just update the pattern to include the other configs, because it appears the actual structure of the schema needs to change depending on which pattern has been selected - i.e. you need something like the following (pseudocode):

```
oneOf
  - name: ".*/appsettings$"
    properties: <AppSettingsProperties>

  - name: ".*/connectionstrings$"
    properties: <ConnectionStringsProperties>
...
```

How has this been handled in previous schemas for your team?

**Comment by anthony-c-martin** [comment]
@pvbchak please could you merge the latest from master and re-run the generation. We've made some fixes to autorest to better handle your constant-named resources.

This should address https://github.com/Azure/azure-resource-manager-schemas/issues/794 which was raised recently.

**Pull request merged**